### PR TITLE
Next Smallest Palindrome

### LINKED ISSUE #886: Next Smallest Palindrome
Eg: Next smallest palindrome of 123 is 131

I want to implement this in c++, java, python and C#.

### PR DESCRIPTION
**Pull request by KanchanThareja**
Fixes #886 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Sample Input / Output.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:

- Implementation of next smallest palindrome of a number.

#### Languages Used:
- c++
- java
- python
- c#

#### Files Added:
-  Next_Smallest_Palindrome.cpp
- Next_Smallest_Palindrome.java
- Next_Smallest_Palindrome.py
- Next_Smallest_Palindrome.cs

### CODE REVIEW

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+        j++; 
+        i--; 
+    } 
+  
+    // the case where middle digit must be incremented.  
+    if (isLeftSmaller) 
+    { 
+        int cflag = 1; 
+        i = middle - 1; 
+        if (n%2 == 1)
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: @KanchanThareja Please add space before and after %

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            i--;
+            j++;
+        }
+
+        // the case where middle digit must be incremented.  
+        if (isLeftSmaller) 
+        { 
+            int cflag = 1; 
+            i = middle - 1; 
+            if (n%2 == 1)
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.java]: @KanchanThareja Add space before and after %

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+        # print the result  
+        for i in range(0, n):  
+            print(int(arr[i]),end=" ")  
+
+    print()   
+
+# Sample input
+num = [9, 4, 1, 8, 7, 9, 7] 
+n = len(num) 
+next_smallest_palindrome( num, n)
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.py]: Remove the space before num

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      
+  
+    # Handling other input cases
+    else: 
+      
+        helper (num, n)  
+  
+        # print the result  
+        for i in range(0, n):  
+            print(int(arr[i]),end=" ")
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.py]: Add space after the comma.

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 5]
```diff
@@ -0,0 +1,106 @@
+#include <stdio.h> 
+
+// This function is for when -> 
+// 1.The input number is not palindrome. 
+// 2.number is palindrome but not all digits as nine
+
+void helper(int num[], int n )
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Please remove space before ).

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+            i--;
+        } 
+    } 
+} 
+
+void next_smallest_palindrome( int num[], int n ) 
+{
+    //if All the digits are 9, o/p 1 followed by n-1 0's followed by 1. 
+    int b = 1;
+    for (int i = 0; i < n; ++i )
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Please remove space before ).

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+            cflag = num[i] / 10; 
+            num[i] %= 10; 
+            num[j] = num[i]; 
+            j++;
+            i--;
+        } 
+    } 
+} 
+
+void next_smallest_palindrome( int num[], int n )
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Please remove space before ) and after (.

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    //if All the digits are 9, o/p 1 followed by n-1 0's followed by 1. 
+    int b = 1;
+    for (int i = 0; i < n; ++i ) 
+        if (num[i] != 9) 
+            b = 0;  
+
+    printf("Next palindrome is:");
+    if (b) 
+    { 
+        printf( "1 ");
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Please remove space after(.

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    int b = 1;
+    for (int i = 0; i < n; ++i ) 
+        if (num[i] != 9) 
+            b = 0;  
+
+    printf("Next palindrome is:");
+    if (b) 
+    { 
+        printf( "1 "); 
+        for(int i = 0; i < n - 1; i++ )
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Please remove space before )

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+    for (int i = 0; i < n; ++i ) 
+        if (num[i] != 9) 
+            b = 0;  
+
+    printf("Next palindrome is:");
+    if (b) 
+    { 
+        printf( "1 "); 
+        for(int i = 0; i < n - 1; i++ ) 
+            printf( "0 " );
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Please remove space after ( and before )

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        if (num[i] != 9) 
+            b = 0;  
+
+    printf("Next palindrome is:");
+    if (b) 
+    { 
+        printf( "1 "); 
+        for(int i = 0; i < n - 1; i++ ) 
+            printf( "0 " ); 
+        printf( "1" );
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.cpp]: Remove space

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.java [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+        // if yes, o/p would be 1 followed by n-1 0's followed by 1. 
+        if(b > 0)
+        {
+            System.out.print(1 + " ");
+
+            for(int j = 0; j < n-1; j++)
+            System.out.print(0 + " ");
+
+            System.out.print(1+ " ");
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.java]: Add space after 1

File: Next_Smallest_Palindrome/Next_Smallest_Palindrome.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    while (i >= 0) :       
+        number[j] = number[i]  
+        j+=1
+        i-=1
+    
+    if (isLeftSmaller == True) :      
+        cflag = 1
+        i = middle - 1
+
+        if (n%2 == 1) :
```
Comment (somya-kapoor) [inline Next_Smallest_Palindrome/Next_Smallest_Palindrome.py]: Add space

**Comment by KanchanThareja** [comment]
@somya-kapoor done

**Comment by KanchanThareja** [comment]
@jainaman224 please review

**Comment by somya-kapoor** [comment]
@jainaman224 This is approved . Please have a final look

**Pull request merged**